cmds/field/field.go

```diff
@@ -0,0 +1,366 @@
+	eflag      = flag.Bool("e", false, "")
+	Eflag      = flag.Bool("E", false, "")
+	insepflag  = flag.String("F", "[ \t\v\r]+", "Input separator characters (regular expression)")
+	outsepflag = flag.String("O", " ", "Output separater (string)")
+)
+
+func main() {
+	flag.Parse()
+
+	fstate := make(map[string]bool)
+	flag.Visit(func(f *flag.Flag) { fstate[f.Name] = true })
+	if fstate["e"] && fstate["E"] {
+		fatal("flag conflict: -e and -E are mutually exclusive")
+	}
+	if fstate["0"] && fstate["O"] {
+		fatal("flag conflict: -O and -0 are mutually exclusive")
+	}
+
+	outsep := *outsepflag
+	insep := *insepflag
+	collapse := shouldcollapse(insep)
+	delim, re_err := regexp.Compile(insep)
+	if re_err != nil {
+		fatal("Delimiter regexp failed to parse: %v", re_err)
+	}
+
+	if flag.NArg() == 0 {
+		fatal("Range specifier missing")
+	}
+	rv := parseranges(flag.Arg(0))
+
+	if flag.NArg() == 1 {
+		process(os.Stdin, rv, delim, outsep, collapse)
+		return
+	}
+	for i := 1; i < flag.NArg(); i++ {
+		filename := flag.Arg(i)
+		if filename == "-" {
+			process(os.Stdin, rv, delim, outsep, collapse)
+			continue
+		}
+		file, err := os.Open(filename)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Cannot open:", err)
+			continue
+		}
+		process(file, rv, delim, outsep, collapse)
+		file.Close()
+	}
+}
+
+func shouldcollapse(s string) bool {
+	if *eflag {
+		return false
+	}
+	if *Eflag {
+		return true
+	}
+	l := utf8.RuneCountInString(s)
+	r, _ := utf8.DecodeRuneInString(s)
+	return l > 1 && (l != 2 || r != '\\')
+}
+
+// The field selection syntax is:
+//
+// ranges := range [[delim] range]
+// range := field | NUM '-' [field]
+// field := NUM | NF
+// delim := ws+ | '|' | ','
+// ws := c such that `isspace(c)` is true.
+// NF := 'NF' | 'N'
+// (Numbers can be negative)
+
+func parseranges(input string) []frange {
+	rs := []frange{}
+	lex := &lexer{input: input}
+	if input == "" {
+		fatal("Empty field range")
+	}
+	lex.next()
+	for {
+		for typ := lex.peektype(); typ == TokSpace; typ = lex.peektype() {
+			lex.next()
+		}
+		r := parserange(lex)
+		rs = append(rs, r)
+		typ := lex.peektype()
+		if typ == TokEOF {
+			break
+		}
+		if !isdelim(typ) {
+			fatal("Syntax error in field list, tok = %s", lex.peektok())
+		}
+		lex.next()
+	}
+	return rs
+}
+
+func parserange(lex *lexer) frange {
+	r := frange{begin: nf, end: nf}
+	if lex.peektype() == TokEOF {
+		fatal("EOF at start of range")
+	}
+	fnum, typ := parsefield(lex)
+	r.begin = fnum
+	r.end = fnum
+	if typ == TokNF {
+		return r
+	}
+	typ = lex.peektype()
+	if typ != TokDash {
+		return r
+	}
+	lex.next()
+	r.end = nf
+	typ = lex.peektype()
+	if typ != TokEOF && !isdelim(typ) {
+		r.end, _ = parsefield(lex)
+	}
+	return r
+}
+
+func parsefield(lex *lexer) (int, toktype) {
+	typ := lex.peektype()
+	if typ == TokNF {
+		lex.next()
+		return nf, TokNF
+	}
+	return parsenum(lex), TokNum
+}
+
+func parsenum(lex *lexer) int {
+	tok, typ := lex.next()
+	if typ == TokEOF {
```

Comment:
```go
func parsenum(lex *lexer) int {
  tok, typ := lex.next()
  switch typ {
  case TokNum:
    num, _ := strconv.Atoi(tok)
    return num
  case TokDash:
    return -1 * parsenum(lex)
  case TokEOF:
  default:
    fatal("negative number parse error: unexpected token '%v', tok)
  }
}
```

slightly different behavior -- feel free to reject, but would prefer switch here either way.

if you want EOF to print for TokEOF, you could define a `fmt.Stringer` on toktype?